tabu/src/bqpSolver.cpp

```diff
@@ -13,80 +13,85 @@
+    double startingObjective,
     int tt,
     long long ZCoeff,
     long long timeLimitInMilliSecs,
     const bqpSolver_Callback *callback
 ) {
     bqp->restartNum++;/*added to record more statistics.*/
     long long startTime = realtime_clock();
-    int i, k, j;
+    int i, k;
     int globalMinFound;
     long /*objectiveChangeCtr = 0,*/ localSearchCtr = 0;
     long long iter = 0;
     int bestK;
-    long localMinCost, cost = 0;
-    long prevCost;
+    double localMinCost, cost = 0;
+    double prevCost;
     vector<int> u_tabu(bqp->nVars);
     int *taboo = vector_data<int>(u_tabu);
     vector<int> u_sol(bqp->nVars);
     int *solution = vector_data<int>(u_sol);
     int tabooTenure = (20 < (int)(bqp->nVars / 4.0))? 20 : (int)(bqp->nVars / 4.0);
     long long maxIter = (500000 > ZCoeff * (long long)bqp->nVars)? 500000 : ZCoeff * (long long)bqp->nVars;
-    vector<long> u_change(bqp->nVars);
-    long *changeInObjective = vector_data<long>(u_change), change;
+    vector<double> u_change(bqp->nVars);
+    double *changeInObjective = vector_data<double>(u_change), change;
 
     vector<int> u_tieList(bqp->nVars);
     int numTies = 0, *tieList = vector_data<int>(u_tieList);
 
-    if (tt > 0) {
+    if (tt > 0) { 
       tabooTenure = tt;
     }
 
-    for(i = 0; i < bqp->nVars; i++) {
+    for (i = 0; i < bqp->nVars; i++) {
         taboo[i] = 0;
         solution[i] = starting[i];
         bqp->solution[i] = starting[i];
         changeInObjective[i] = bqpUtil_getChangeInObjective(bqp, starting, i);
     }
     bqp->solutionQuality = startingObjective;
     prevCost = bqp->solutionQuality;
-    while(iter < maxIter && (realtime_clock() - startTime) < timeLimitInMilliSecs) {
+
+    while (iter < maxIter && (realtime_clock() - startTime) < timeLimitInMilliSecs) {
         bqp->iterNum++; /*added to record more statistics.*/
-        localMinCost = LARGE_NUMBER;
+        localMinCost = std::numeric_limits<double>::max();
         bestK = -1;
         globalMinFound = 0;
         numTies = 0;
         bqp->evalNum += bqp->nVars; /*added to record more statistics.*/
-        for(k = 0; k < bqp->nVars; k++) {
-            if(taboo[k] != 0) {
+
+        for (k = 0; k < bqp->nVars; k++) {
+            if (taboo[k] != 0) {
                 continue;
             }
             iter++;
             cost = prevCost + changeInObjective[k];
-            if(cost < bqp->solutionQuality) {
+            if (cost < bqp->solutionQuality) {
+                // A better solution was found. Local search procedure will be invoked later.
                 bestK = k;
                 globalMinFound = 1;
                 break;
             }
-            if(cost < localMinCost) {
+            if (cost < localMinCost) {
                 bestK = k;
                 localMinCost = cost;
                 tieList[0] = k;
                 numTies = 1;
             }
-            else if(cost == localMinCost) {
+            else if (cost == localMinCost) {
```

Comment:
I think you could do an approximate equality check here, which may help in the case where states which have essentially the same objective but look different because of float errors (though I would ask... how is the objective function actually defined on floats? float addition is not associative :stuck_out_tongue: ). However, none of the other solvers we have do this kind of check to my knowledge, so I don't see a strong reason to do it here.